File: frontend/src/services/intentRouter.js
// frontend/src/services/intentRouter.js

/**
 * Servicio para procesamiento de intenciones de voz
 * Implementa QGAI-107: Router de intenciones con fallback
 */

import { logIntentEvent } from "./voiceMetricsService";
const API_BASE = process.env.REACT_APP_API_BASE || "/api";


class IntentRouterService {
  constructor() {
    this.baseUrl = `${API_BASE}/intent-router`;
    this.cache = new Map();
    this.cacheTimeout = 5 * 60 * 1000; // 5 minutos
  }

  // Helper interno para llamadas fetch con JSON y manejo de errores
  async _fetch(path, { method = 'GET', data } = {}) {
    const init = {
      method,
      headers: { 'Content-Type': 'application/json' },
    };
    if (data !== undefined) init.body = JSON.stringify(data);

    const resp = await fetch(`${this.baseUrl}${path}`, init);
    let json = null;
    try { json = await resp.json(); } catch {}
    if (!resp.ok) {
      const msg = (json && (json.detail || json.error || json.message)) || `HTTP ${resp.status}`;
      const err = new Error(msg);
      err.status = resp.status;
      err.payload = json;
      throw err;
    }
    return json ?? {};
  }

  /**
   * Parsea un comando de voz y retorna la intención
   * @param {string} text - Texto transcrito del comando
   * @returns {Promise<IntentResult>}
   */
  async parseIntent(text) {
    if (!text || text.trim().length === 0) {
      return this._createErrorResult('Texto vacío');
    }

    // Check cache primero
    const cacheKey = text.toLowerCase().trim();
    const cached = this.cache.get(cacheKey);
    if (cached && Date.now() - cached.timestamp < this.cacheTimeout) {
      return cached.result;
    }

    try {
      const response = await this._fetch('/parse/', {
        method: 'POST',
        data: { text }
      });

      const result = {
        intent: response.intent,
        confidence: response.confidence,
        slots: response.slots || {},
        backendUsed: response.backend_used,
        latencyMs: response.latency_ms,
        warning: response.warning || null,
        timestamp: Date.now()
      };

      // log de métricas (no bloquea UX)
      logIntentEvent(
        result.intent,
        Number(result.confidence ?? 0),
        String(result.backendUsed || "grammar"),
        Number(result.latencyMs ?? 0),
        result.slots || {},
        { source: "frontend" }
      ).catch(()=>{});

      // Cache result
      this.cache.set(cacheKey, {
        result,
        timestamp: Date.now()
      });

      // Analytics event
      this._trackIntent(result);

      return result;

    } catch (error) {
      console.error('Intent parsing error:', error);
      // Fallback local básico
      return this._localFallback(text);
    }
  }

  /**
   * Parsea múltiples comandos en batch
   * @param {string[]} texts - Array de textos
   * @returns {Promise<IntentResult[]>}
   */
  async parseBatch(texts) {
    if (!Array.isArray(texts) || texts.length === 0) {
      return [];
    }

    try {
      const response = await this._fetch('/batch_parse/', {
        method: 'POST',
        data: { texts }
      });

      return (response.results || []).map(r => ({
        text: r.text,
        intent: r.intent,
        confidence: r.confidence,
        slots: r.slots || {},
        backendUsed: r.backend_used,
        latencyMs: r.latency_ms
      }));

    } catch (error) {
      console.error('Batch parsing error:', error);
      // Fallback: parsear uno por uno localmente
      return texts.map(text => this._localFallback(text));
    }
  }

  /**
   * Obtiene lista de intents soportados
   * @returns {Promise<Object>}
   */
  async getSupportedIntents() {
    try {
      const response = await this._fetch('/supported_intents/');
      return response;
    } catch (error) {
      console.error('Error fetching supported intents:', error);
      return { intents: {} };
    }
  }

  /**
   * Verifica salud de los backends
   * @returns {Promise<Object>}
   */
  async checkHealth() {
    try {
      const response = await this._fetch('/health/');
      return response;
    } catch (error) {
      console.error('Health check error:', error);
      return {
        status: 'error',
        backends: {
          grammar: 'unknown',
          gemini: 'unknown',
          perplexity: 'unknown'
        }
      };
    }
  }

  /**
   * Fallback local usando patrones simples
   * @private
   */
  _localFallback(text) {
    const textLower = text.toLowerCase().trim();

    // Patrones básicos locales (más sinónimos y variantes)
    // Nota: el orden prioriza patrones más específicos primero
    const patterns = {
      // Destructivas / estructurales
      delete_question: /\b(eliminar|borra?r?|borrar|quitar|suprimir|sacar|remover|quita)\b/,
      duplicate_question: /\b(duplicar|clonar|copiar|haz\s+una\s+copia|duplicame|duplicar\s+pregunta)\b/,

      // Export / descargar
      export_pdf: /\b(exportar|exporta|descargar|guardar|generar|imprimir|bajar)\b.*\b(pdf|como pdf|en pdf)\b/,
      export_txt: /\b(exportar|exporta|descargar|guardar|bajar)\b.*\b(txt|texto|txt|archivo de texto)\b/,
      export_quiz: /\b(exportar|exporta|descargar|guardar|bajar|export)\b/,

      // Regenerar / reemplazar
      regenerate_quiz: /\b(regenerar|regenera|vuelve?\s+a\s+generar|volver\s+a\s+crear|reemplazar|sustituir|renovar)\b/,

      // Navegación
      navigate_next: /\b(siguiente|próxima?|continua?r?|adelante|next|avanza|sigue|pasar|pasa)\b/,
      navigate_previous: /\b(anterior|atr[aá]s|volver|back|retrocede|regresa)\b/,

      // Interacción con pregunta
      read_question: /\b(lee?r?|leer|pregunta|leer\s+la\s+pregunta|leer\s+pregunta)\b/,
      show_answers: /\b(muestra?r?|mostrar|ver|respuestas?|opciones|ensena|enséñame)\b/,

      // Generación de quiz
      generate_quiz: /\b(genera?r?|crea?r?|arma?r?|haz|hazme|genera|crea|arma|quiz|cuestionario|test|preguntas)\b/,

      // Repetir / pausa / reanudar
      repeat: /\b(repite?|repetir|otra\s+vez|de\s+nuevo|otra)\b/,
      pause: /\b(pausa?|pausar|detener|detene?r?|stop|para)\b/,
      resume: /\b(continua?r?|reanuda?r?|resume|seguir)\b/,
      skip: /\b(salta?r?|omitir|skip|siguiente)\b/,
      finish: /\b(terminar|finalizar|salir|finish|acabar)\b/,
      slower: /\b(lento|despacio|slower|mas lento|más lento)\b/
    };

    for (const [intent, pattern] of Object.entries(patterns)) {
      if (pattern.test(textLower)) {
        // Extraído simple de slots: número y tema/dificultad para generación
        const slots = {};
        // número: "10 preguntas", "diez preguntas"
        const numMatch = textLower.match(/(\d+)|uno|dos|tres|cuatro|cinco|seis|siete|ocho|nueve|diez/);
        if (numMatch) {
          let n = numMatch[0];
          // convetir texto a número si es palabra (muy básico)
          const numWords = { uno:1,dos:2,tres:3,cuatro:4,cinco:5,seis:6,siete:7,ocho:8,nueve:9,diez:10 };
          if (isNaN(Number(n))) n = numWords[n] || null;
          slots.count = n ? Number(n) : undefined;
        }
        // tema: "sobre geografía", "de historia"
        const temaMatch = textLower.match(/(?:sobre|de|acerca de)\s+([a-záéíóúñ0-9\s]+)/);
        if (temaMatch) slots.topic = (temaMatch[1] || '').trim();
        // dificultad: normalizar a 'Fácil' | 'Media' | 'Difícil'
        const difMatch = textLower.match(/\b(f(a|á)cil|facil|sencillo|baja\s+dificultad|bajo)\b|\b(medi[oa]|intermedio|intermedia|normal|regular|moderad[oa]|nivel\s+medio)\b|\b(dif(i|í)cil|dificil|complicad[oa]|duro|avanzad[oa]|alta\s+dificultad|experto|profundo)\b/);
        if (difMatch) {
          const hit = difMatch[0];
          if (/\b(f(a|á)cil|facil|sencillo|baja\s+dificultad|bajo)\b/.test(hit)) slots.difficulty = 'Fácil';
          else if (/\b(medi[oa]|intermedio|intermedia|normal|regular|moderad[oa]|nivel\s+medio)\b/.test(hit)) slots.difficulty = 'Media';
          else slots.difficulty = 'Difícil';
        }

        return {
          intent,
          confidence: 0.6,
          slots,
          backendUsed: 'local_fallback',
          latencyMs: 0,
          warning: 'Using local fallback'
        };
      }
    }

    return {
      intent: 'unknown',
      confidence: 0.0,
      slots: {},
      backendUsed: 'local_fallback',
      latencyMs: 0,
      warning: 'Intent not recognized'
    };
  }

  /**
   * Crea un resultado de error
   * @private
   */
  _createErrorResult(message) {
    return {
      intent: 'unknown',
      confidence: 0.0,
      slots: {},
      backendUsed: 'error',
      latencyMs: 0,
      warning: message
    };
  }

  /**
   * Trackea evento de intención en analytics
   * @private
   */
  _trackIntent(result) {
    // Integración con analytics (PostHog, GA4, etc.)
    if (window.analytics) {
      window.analytics.track('Intent Recognized', {
        intent: result.intent,
        confidence: result.confidence,
        backend: result.backendUsed,
        latency_ms: result.latencyMs,
        has_warning: !!result.warning
      });
    }
  }

  /**
   * Limpia la caché
   */
  clearCache() {
    this.cache.clear();
  }
}

// Singleton instance
const intentRouter = new IntentRouterService();

export default intentRouter;

File: backend/api/services/intent_router.py
# backend/api/services/intent_router.py

import re
import time
import logging
from typing import Dict, Optional, Tuple
from dataclasses import dataclass
from enum import Enum
import google.generativeai as genai
from openai import OpenAI
import requests
from django.conf import settings
from django.core.cache import cache

logger = logging.getLogger(__name__)


class IntentBackend(Enum):
    GRAMMAR = "grammar"
    GEMINI = "gemini"
    PERPLEXITY = "perplexity"
    FALLBACK = "fallback"


@dataclass
class IntentResult:
    intent: str
    confidence: float
    slots: Dict
    backend_used: IntentBackend
    latency_ms: float
    raw_text: str = ""
    error: Optional[str] = None


class IntentGrammar:
    """Sistema de gramática/regex para comandos básicos"""
    
    INTENT_PATTERNS = {
        'generate_quiz': [
            r'(?:genera?|crea?|arma|haz|hazme)\s+(?:un\s+)?(?:quiz|cuestionario|examen|test)?\s*(?:de|sobre|acerca\s+de)?\s+(.+)',
            r'(?:quiz|cuestionario|test)\s+(?:de|sobre)\s+(.+)',
        ],
        'read_question': [
            r'(?:lee?|leeme|leer)\s+(?:la\s+)?pregunta\s+(\d+)',
            r'pregunta\s+(?:número\s+)?(\d+)',
        ],
        'navigate_next': [
            r'(?:siguiente|próxima?|adelante|continua?r?|next)',
        ],
        'navigate_previous': [
            r'(?:anterior|atrás|volver|back)',
        ],
        'show_answers': [
            r'(?:muestra?|mostrar|ver|enseña?r?)\s+(?:las\s+)?(?:respuestas?|opciones)',
        ],
        'regenerate': [
            r'(?:regenera?r?|vuelve\s+a\s+generar)\s+(?:la\s+)?pregunta\s+(\d+)',
            r'(?:regenera?r?|cambiar)\s+(?:pregunta\s+)?(\d+)',
        ],
        'export': [
            r'(?:exporta?r?|descargar|guardar)\s+(?:el\s+)?(?:quiz|cuestionario|test)?',
        ],
        'repeat': [
            r'(?:repite?|repetir|otra\s+vez|de\s+nuevo)',
        ],
        'slower': [
            r'(?:más\s+lento|despacio|slower)',
        ],
        'pause': [
            r'(?:pausa?|pausar|detene?r?|stop)',
        ],
        'resume': [
            r'(?:continua?r?|reanuda?r?|resume)',
        ],
        'skip': [
            r'(?:salta?r?|omitir|skip)',
        ],
        'finish': [
            r'(?:terminar|finalizar|salir|finish)',
        ],
    }
    
    # Mapeo de sinónimos para difficulty
    DIFFICULTY_MAP = {
        'fácil': 'easy',
        'facil': 'easy',
        'sencillo': 'easy',
        'sencilla': 'easy',
        'simple': 'easy',
        'básico': 'easy',
        'basico': 'easy',
        'medio': 'medium',
        'media': 'medium',
        'intermedio': 'medium',
        'intermedia': 'medium',
        'normal': 'medium',
        'difícil': 'hard',
        'dificil': 'hard',
        'complicado': 'hard',
        'complicada': 'hard',
        'avanzado': 'hard',
        'avanzada': 'hard',
    }
    
    @classmethod
    def match(cls, text: str) -> Optional[Tuple[str, Dict, float]]:
        """
        Intenta hacer match con gramática/regex
        Returns: (intent, slots, confidence) o None
        """
        text_lower = text.lower().strip()
        
        for intent, patterns in cls.INTENT_PATTERNS.items():
            for pattern in patterns:
                match = re.search(pattern, text_lower, re.IGNORECASE)
                if match:
                    slots = cls._extract_slots(intent, match, text_lower)
                    confidence = cls._calculate_confidence(intent, text_lower)
                    return intent, slots, confidence
        
        return None
    
    @classmethod
    def _extract_slots(cls, intent: str, match, text: str) -> Dict:
        """Extrae slots del match"""
        slots = {}
        
        if intent == 'generate_quiz':
            # Extraer tema
            slots['topic'] = match.group(1).strip()
            
            # Buscar dificultad
            for key, value in cls.DIFFICULTY_MAP.items():
                if key in text:
                    slots['difficulty'] = value
                    break
            
            # Buscar número de preguntas
            num_match = re.search(r'(\d+)\s+(?:preguntas?|items?)', text)
            if num_match:
                slots['num_questions'] = int(num_match.group(1))
        
        elif intent in ['read_question', 'regenerate']:
            if match.groups():
                slots['question_number'] = int(match.group(1))
        
        return slots
    
    @classmethod
    def _calculate_confidence(cls, intent: str, text: str) -> float:
        """Calcula score de confianza basado en palabras clave"""
        base_confidence = 0.85
        
        # Palabras clave que aumentan confianza
        keyword_boost = {
            'generate_quiz': ['genera', 'crea', 'quiz', 'cuestionario'],
            'read_question': ['lee', 'leer', 'pregunta'],
            'navigate_next': ['siguiente', 'próxima'],
        }
        
        if intent in keyword_boost:
            matches = sum(1 for word in keyword_boost[intent] if word in text)
            boost = min(matches * 0.05, 0.15)
            return min(base_confidence + boost, 1.0)
        
        return base_confidence


class GeminiNLU:
    """Cliente para Gemini API"""
    
    def __init__(self):
        self.api_key = settings.GEMINI_API_KEY
        genai.configure(api_key=self.api_key)
        self.model = genai.GenerativeModel('gemini-pro')
        self.timeout = 3.0  # 3 segundos
        
        self.system_prompt = """Eres un clasificador de intenciones para una app de quiz educativa.
Analiza el comando de voz del usuario y devuelve SOLO un JSON con este formato:
{
  "intent": "nombre_del_intent",
  "confidence": 0.95,
  "slots": {"param": "valor"}
}

Intents válidos:
- generate_quiz: crear quiz (slots: topic, difficulty, num_questions)
- read_question: leer pregunta (slots: question_number)
- navigate_next: siguiente pregunta
- navigate_previous: pregunta anterior
- show_answers: mostrar respuestas
- regenerate: regenerar pregunta (slots: question_number)
- export: exportar quiz
- repeat: repetir última acción
- slower: hablar más lento
- pause: pausar
- resume: reanudar
- skip: saltar pregunta
- finish: terminar sesión
- unknown: no se entiende

SOLO devuelve el JSON, sin explicaciones."""
    
    def parse_intent(self, text: str) -> Tuple[str, Dict, float]:
        """Parsea intención usando Gemini"""
        try:
            prompt = f"{self.system_prompt}\n\nComando del usuario: \"{text}\""
            
            response = self.model.generate_content(
                prompt,
                generation_config=genai.types.GenerationConfig(
                    temperature=0.1,
                    max_output_tokens=200,
                )
            )
            
            # Parsear respuesta JSON
            import json
            result_text = response.text.strip()
            
            # Limpiar markdown si existe
            if result_text.startswith('```'):
                result_text = result_text.split('```')[1]
                if result_text.startswith('json'):
                    result_text = result_text[4:]
                result_text = result_text.strip()
            
            result = json.loads(result_text)
            
            return (
                result.get('intent', 'unknown'),
                result.get('slots', {}),
                result.get('confidence', 0.5)
            )
            
        except Exception as e:
            logger.error(f"Gemini parsing error: {e}")
            raise


class PerplexityNLU:
    """Cliente para Perplexity API"""
    
    def __init__(self):
        self.api_key = settings.PERPLEXITY_API_KEY
        self.client = OpenAI(
            api_key=self.api_key,
            base_url="https://api.perplexity.ai"
        )
        self.timeout = 4.0  # 4 segundos
        
        self.system_prompt = """You are an intent classifier for an educational quiz app.
Analyze the voice command and return ONLY a JSON with this format:
{
  "intent": "intent_name",
  "confidence": 0.95,
  "slots": {"param": "value"}
}

Valid intents: generate_quiz, read_question, navigate_next, navigate_previous, 
show_answers, regenerate, export, repeat, slower, pause, resume, skip, finish, unknown

Return ONLY the JSON, no explanations."""
    
    def parse_intent(self, text: str) -> Tuple[str, Dict, float]:
        """Parsea intención usando Perplexity"""
        try:
            response = self.client.chat.completions.create(
                model="llama-3.1-sonar-small-128k-online",
                messages=[
                    {"role": "system", "content": self.system_prompt},
                    {"role": "user", "content": f"Command: \"{text}\""}
                ],
                temperature=0.1,
                max_tokens=200,
                timeout=self.timeout
            )
            
            # Parsear respuesta
            import json
            result_text = response.choices[0].message.content.strip()
            
            # Limpiar markdown
            if result_text.startswith('```'):
                result_text = result_text.split('```')[1]
                if result_text.startswith('json'):
                    result_text = result_text[4:]
                result_text = result_text.strip()
            
            result = json.loads(result_text)
            
            return (
                result.get('intent', 'unknown'),
                result.get('slots', {}),
                result.get('confidence', 0.5)
            )
            
        except Exception as e:
            logger.error(f"Perplexity parsing error: {e}")
            raise


class IntentRouter:
    """Router principal con fallback en cascada"""
    
    def __init__(self):
        self.grammar = IntentGrammar()
        self.gemini = None
        self.perplexity = None
        
        # Feature flags from settings/env
        self.use_gemini = getattr(settings, 'ENABLE_GEMINI_NLU', True)
        self.use_perplexity = getattr(settings, 'ENABLE_PERPLEXITY_NLU', True)
        
        # Umbrales
        self.grammar_threshold = 0.7
        self.llm_threshold = 0.6
        
        # Rate limiting
        self.rate_limit_key_prefix = "intent_router_rate_limit"
        self.rate_limit_per_user = 100  # por hora
    
    def _check_rate_limit(self, user_id: str) -> bool:
        """Verifica rate limit por usuario"""
        key = f"{self.rate_limit_key_prefix}:{user_id}"
        current = cache.get(key, 0)
        
        if current >= self.rate_limit_per_user:
            return False
        
        cache.set(key, current + 1, 3600)  # 1 hora
        return True
    
    def route(self, text: str, user_id: str) -> IntentResult:
        """
        Pipeline principal: Grammar → Gemini → Perplexity → Fallback
        """
        start_time = time.time()
        
        # Rate limiting
        if not self._check_rate_limit(user_id):
            logger.warning(f"Rate limit exceeded for user {user_id}")
            return IntentResult(
                intent='unknown',
                confidence=0.0,
                slots={},
                backend_used=IntentBackend.FALLBACK,
                latency_ms=0,
                error="Rate limit exceeded"
            )
        
        # 1. Intentar con gramática primero
        grammar_result = self.grammar.match(text)
        if grammar_result:
            intent, slots, confidence = grammar_result
            
            if confidence >= self.grammar_threshold:
                latency = (time.time() - start_time) * 1000
                logger.info(f"Grammar match: {intent} (conf: {confidence})")
                
                return IntentResult(
                    intent=intent,
                    confidence=confidence,
                    slots=slots,
                    backend_used=IntentBackend.GRAMMAR,
                    latency_ms=latency,
                    raw_text=text
                )
        
        # 2. Intentar con Gemini
        if self.use_gemini:
            try:
                if not self.gemini:
                    self.gemini = GeminiNLU()
                
                gemini_start = time.time()
                intent, slots, confidence = self.gemini.parse_intent(text)
                gemini_latency = (time.time() - gemini_start) * 1000
                
                if gemini_latency > 3000:  # Timeout
                    logger.warning(f"Gemini timeout: {gemini_latency}ms")
                    raise TimeoutError("Gemini exceeded timeout")
                
                if confidence >= self.llm_threshold:
                    latency = (time.time() - start_time) * 1000
                    logger.info(f"Gemini match: {intent} (conf: {confidence}, latency: {gemini_latency}ms)")
                    
                    return IntentResult(
                        intent=intent,
                        confidence=confidence,
                        slots=slots,
                        backend_used=IntentBackend.GEMINI,
                        latency_ms=latency,
                        raw_text=text
                    )
                    
            except Exception as e:
                logger.error(f"Gemini failed: {e}")
        
        # 3. Fallback a Perplexity
        if self.use_perplexity:
            try:
                if not self.perplexity:
                    self.perplexity = PerplexityNLU()
                
                perplexity_start = time.time()
                intent, slots, confidence = self.perplexity.parse_intent(text)
                perplexity_latency = (time.time() - perplexity_start) * 1000
                
                latency = (time.time() - start_time) * 1000
                logger.info(f"Perplexity match: {intent} (conf: {confidence}, latency: {perplexity_latency}ms)")
                
                return IntentResult(
                    intent=intent,
                    confidence=confidence,
                    slots=slots,
                    backend_used=IntentBackend.PERPLEXITY,
                    latency_ms=latency,
                    raw_text=text
                )
                
            except Exception as e:
                logger.error(f"Perplexity failed: {e}")
        
        # 4. Fallback final - comandos básicos
        latency = (time.time() - start_time) * 1000
        fallback_intent = self._fallback_intent(text)
        
        logger.warning(f"All backends failed, using fallback: {fallback_intent}")
        
        return IntentResult(
            intent=fallback_intent,
            confidence=0.3,
            slots={},
            backend_used=IntentBackend.FALLBACK,
            latency_ms=latency,
            raw_text=text,
            error="All backends failed"
        )
    
    def _fallback_intent(self, text: str) -> str:
        """Fallback simple basado en palabras clave"""
        text_lower = text.lower()
        
        # Comandos más comunes
        if any(word in text_lower for word in ['siguiente', 'próxima', 'next']):
            return 'navigate_next'
        elif any(word in text_lower for word in ['anterior', 'atrás', 'back']):
            return 'navigate_previous'
        elif any(word in text_lower for word in ['genera', 'crea', 'haz']):
            return 'generate_quiz'
        elif any(word in text_lower for word in ['lee', 'leer', 'read']):
            return 'read_question'
        elif any(word in text_lower for word in ['repite', 'repetir', 'repeat']):
            return 'repeat'
        elif any(word in text_lower for word in ['pausa', 'detén', 'stop']):
            return 'pause'
        
        return 'unknown'